panic in master

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x9ec9ee]

goroutine 4064 [running]:
io/ioutil.readAll.func1(0xc011a5bae0)
	/usr/local/go/src/io/ioutil/ioutil.go:30 +0x105
panic(0xebd0a0, 0x197eaa0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/grafana/metrictank/cluster.(*RawResp).Read(0xc003d98a60, 0xc011a58a00, 0x200, 0x200, 0x200, 0x200, 0xc011a58a00)
	<autogenerated>:1 +0x2e
bytes.(*Buffer).ReadFrom(0xc00022f810, 0x7fde84619b50, 0xc003d98a60, 0xc00001c000, 0xed3020, 0xf51c20)
	/usr/local/go/src/bytes/buffer.go:206 +0xbd
io/ioutil.readAll(0x7fde84619b50, 0xc003d98a60, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/ioutil/ioutil.go:36 +0xb5
io/ioutil.ReadAll(0x7fde84619b50, 0xc003d98a60, 0xc003d98a60, 0x7fde84619b50, 0xc003d98a60, 0x25d0e33d, 0xed60a753a)
	/usr/local/go/src/io/ioutil/ioutil.go:45 +0x3e
github.com/grafana/metrictank/cluster.HTTPNode.Post(0xc0000a87e0, 0x2d, 0xc0001b4be0, 0x14, 0x0, 0x25d0e33d, 0xed60a753a, 0x0, 0x0, 0x1, ...)
	/go/src/github.com/grafana/metrictank/cluster/node.go:344 +0x1e0
github.com/grafana/metrictank/api.(*Server).queryAllShards.func1(0x117d980, 0xc000170ac0, 0x1184640, 0xc0001e3590, 0x1000000000000, 0x0, 0x199b9c0, 0xc0006eeec8)
	/go/src/github.com/grafana/metrictank/api/cluster.go:482 +0x9a
github.com/grafana/metrictank/api.queryPeers.func1.1(0xc000000008, 0x1184640, 0xc0001e3590)
	/go/src/github.com/grafana/metrictank/api/cluster.go:545 +0x7a
created by github.com/grafana/metrictank/api.queryPeers.func1
	/go/src/github.com/grafana/metrictank/api/cluster.go:564 +0x391

```
I think what happens here is that because we now use an io.ReadCloser to be a wrapper type such that we can use it to close the span, we can't easily check if the body is non-nil anymore.
in particular the case where error is nil but the body is also nil, can't currently be detected.

--- a/api/dataprocessor.go
+++ b/api/dataprocessor.go
@@ -256,11 +256,11 @@ func (s *Server) getTargetsRemote(ctx context.Context, ss *models.StorageStats, 
 			return resp, nil
 		}
 		body, err := node.PostRaw(rCtx, "getTargetsRemote", "/getdata", models.GetData{Requests: reqs})
-		defer body.Close()
-		if err != nil {
+		if body == nil || err != nil {
 			return nil, err
 		}
 		err = msgp.Decode(body, &resp)
+		body.Close()
 		return resp, err
 	})
 

--- a/api/graphite.go
+++ b/api/graphite.go
@@ -1031,11 +1031,11 @@ func (s *Server) clusterFindByTag(ctx context.Context, orgId uint32, expressions
 		func(reqCtx context.Context, peer cluster.Node) (interface{}, error) {
 			resp := models.IndexFindByTagResp{}
 			body, err := peer.PostRaw(reqCtx, "clusterFindByTag", "/index/find_by_tag", data)
-			defer body.Close()
-			if err != nil {
+			if body == nil || err != nil {
 				return nil, err
 			}
 			err = msgp.Decode(body, &resp)
+			body.Close()
 			return resp, err
 		})
 

--- a/cluster/if.go
+++ b/cluster/if.go
@@ -12,7 +12,6 @@ type Node interface {
 	GetPriority() int
 	HasData() bool
 	Post(context.Context, string, string, Traceable) ([]byte, error)
-	// the returned ReadCloser is always non-nil and Close() must always be called on it, regardless of error state
 	PostRaw(ctx context.Context, name, path string, body Traceable) (io.ReadCloser, error)
 	GetName() string
 }

--- a/cluster/node.go
+++ b/cluster/node.go
@@ -255,21 +255,6 @@ func (n *HTTPNode) SetPartitions(part []int32) {
 	n.Updated = time.Now()
 }
 
-type RawResp struct {
-	io.ReadCloser
-	span opentracing.Span
-}
-
-func (rr RawResp) Close() error {
-	rr.span.Finish()
-	if rr.ReadCloser != nil {
-		return rr.ReadCloser.Close()
-	}
-	return nil
-}
-
-// PostRaw executes a post request on the HTTPNode
-// Important: Close() must always be called on the returned ReadCloser, regardless of error state!
 func (n HTTPNode) PostRaw(ctx context.Context, name, path string, body Traceable) (io.ReadCloser, error) {
 	ctx, span := tracing.NewSpan(ctx, Tracer, name)
 	tags.SpanKindRPCClient.Set(span)
@@ -285,21 +270,18 @@ func (n HTTPNode) PostRaw(ctx context.Context, name, path string, body Traceable
 		if err != nil || time.Since(pre) > 10*time.Second {
 			body.TraceDebug(span)
 		}
+		span.Finish()
 	}(time.Now())
-
-	resp := RawResp{
-		span: span,
-	}
 
 	b, err := json.Marshal(body)
 	if err != nil {
-		return resp, NewError(http.StatusInternalServerError, err)
+		return nil, NewError(http.StatusInternalServerError, err)
 	}
 	reader := bytes.NewReader(b)
 	addr := n.RemoteURL() + path
 	req, err := http.NewRequest("POST", addr, reader)
 	if err != nil {
-		return resp, NewError(http.StatusInternalServerError, err)
+		return nil, NewError(http.StatusInternalServerError, err)
 	}
 	req = req.WithContext(ctx)
 	carrier := opentracing.HTTPHeadersCarrier(req.Header)
@@ -316,31 +298,30 @@ func (n HTTPNode) PostRaw(ctx context.Context, name, path string, body Traceable
 	case <-ctx.Done():
 		log.Debugf("CLU HTTPNode: context canceled on request to peer %s", n.Name)
 		err = nil
-		return resp, nil
+		return nil, nil
 	default:
 	}
 
 	if err != nil {
 		tags.Error.Set(span, true)
 		log.Errorf("CLU HTTPNode: error trying to talk to peer %s: %s", n.Name, err.Error())
-		return resp, NewError(http.StatusServiceUnavailable, errors.New("error trying to talk to peer"))
+		return nil, NewError(http.StatusServiceUnavailable, errors.New("error trying to talk to peer"))
 	}
 	if rsp.StatusCode != 200 {
 		// Read in body so that the connection can be reused
 		io.Copy(ioutil.Discard, rsp.Body)
 		rsp.Body.Close()
-		return resp, NewError(rsp.StatusCode, fmt.Errorf(rsp.Status))
+		return nil, NewError(rsp.StatusCode, fmt.Errorf(rsp.Status))
 	}
-	resp.ReadCloser = rsp.Body
-	return resp, nil
+	return rsp.Body, nil
 }
 
 func (n HTTPNode) Post(ctx context.Context, name, path string, body Traceable) (ret []byte, err error) {
 	bodyReader, err := n.PostRaw(ctx, name, path, body)
-	defer bodyReader.Close()
-	if err != nil {
+	if err != nil || bodyReader == nil {
 		return nil, err
 	}
+	defer bodyReader.Close()
 	return ioutil.ReadAll(bodyReader)
 }
 

